V2

albertwoo opened the pull request:
1. Removed dependency on bolero. I implemented EltBuilder and ComponentBuilder to have better support CE style. With this we can directly access blazor internal render tree builder to build dom tree with better performance, also with better and flexiable DSL.
2. Feliz style is not recommend way now. Even it's performance is better than V1, but it also make the cli generator not easy to make type safe. Also it is too verbose. CE style is cleaner, even the intelicense is not that good, but in most cases, we only need to give a hint for want we want and make sure the compile time is safe.
3. I also plan to remove elmish out from the Fun.Blazor to a seperate package. So it can be lighter for people who do not want use elmish. Currently it is not ported. Will port it in the future commit.
4. The initial idea is from [bolero experiment to use f# inline feature](https://github.com/fsbolero/Bolero/issues/249), but it turns out very big bundle size because too much duplication code. So I removed a lot of that improvement. So the final generated code may not be as beautiful as I expected to be but at least I removed all the unnecessary loops.

Breaking changes: for all the internal CE style element like div, span, we no need to construct it anymore
```fsharp
div() {
    ...
}
```
should be changed to 
```fsharp
div {
    ...
}
```

Breaking changes: we cannot mix the order of attribute and child content anymore. All the attributes should be write before the child content.. Below code will not compile:
```fsharp
div {
    "Wrong way"
    styleBuilder { color "red" }
}
```
should be changed to 
```fsharp
div {
    styleBuilder { color "red" }
    "Wrong way"
}
```

Recommendation:
1. We should reduce the use of list or sequence to build dom tree for better performance (less loop, less allocation)
```fsharp
div {
    childContent [
        div
        div
    ]
}
```
is recommend to change to
```fsharp
div {
    childContent (fragment {
        div
        div
    })
}
```
or just
```fsharp
div {
    div
    div
}
```

AngelMunoz commented:
Does this mean we won't be able to interop nicely with bolero components/libraries?
or do we have to do it in a different way? perhaps using FunBlazorComponent maybe or providing a different package with a Bolero Node transformer


The main reasoning I have in mind is due how Fable.React and Feliz DSLs can interop between each other and reduce the amount of annoyances and barriers within the ecosystem itself.

Maybe not on this release but perhaps it could be revisited later

AngelMunoz commented:
By the way amazing work here!

albertwoo commented:
@AngelMunoz interop with bolero into Fun.Blazor is very easy. But vice versa will need a helper function. You can check https://github.com/slaveOftime/Fun.Blazor/blob/V2/Fun.Blazor.Docs.Wasm/Pages/HelperFunctions/InteropWithBoleroDemo.fs

albertwoo commented:
Will merge this to master and generate first beta version to test more.

granicz commented:
@albertwoo Can you elaborate on item 4) above about the increased bundle size and what changes you made to alleviate that?

albertwoo commented:
@granicz I always use the docs app to check. And found it from around 4MB to 8MB. So what I changed is that for internal element like div etc, I will use inline as more as possible. But for third party library generator cli, I will not add that.

The bundle size increased a lot, also because I added multiple third party libraries like MudBlazor, AntDesgin etc... So I am planning to give an option for the cli to toggle if inline should be used. Because most of the time, they only have one main third party UI library.

Also if the trim is possible, then the bundle size can also be reduced.
